File: server/src/main/java/cn/enaium/blog/controller/PostController.java
package cn.enaium.blog.controller;

import cn.dev33.satoken.stp.StpUtil;
import cn.enaium.blog.model.entity.CategoryFetcher;
import cn.enaium.blog.model.entity.Post;
import cn.enaium.blog.model.entity.PostDraft;
import cn.enaium.blog.model.entity.PostFetcher;
import cn.enaium.blog.model.entity.input.PostInput;
import cn.enaium.blog.repository.PostRepository;
import lombok.AllArgsConstructor;
import org.babyfish.jimmer.client.FetchBy;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.UUID;

/**
 * @author Enaium
 */
@RestController
@AllArgsConstructor
public class PostController {

    private final PostRepository repository;

    @PutMapping("/posts/")
    @ResponseStatus(HttpStatus.OK)
    public void createPost(@RequestBody PostInput postInput) {
        repository.insert(PostDraft.$.produce(draft -> {
            draft.setUserId(UUID.fromString(StpUtil.getLoginIdAsString()));
            draft.setTitle(postInput.getTitle());
            draft.setContent(postInput.getContent());
            draft.setCategoryId(postInput.getCategoryId());
        }));
    }

    @GetMapping("/posts/")
    public Page<@FetchBy("DEFAULT_POST") Post> findPosts(@RequestParam(defaultValue = "0") Integer page, @RequestParam(defaultValue = "10") Integer size) {
        return repository.findAll(PageRequest.of(page, size), DEFAULT_POST);
    }

    @GetMapping("/posts/{id}")
    public @FetchBy("FULL_POST") Post findPost(@PathVariable UUID id) {
        return repository.findNullable(id, FULL_POST);
    }

    @GetMapping("/categories/{categoryId}/posts/")
    public Page<@FetchBy("DEFAULT_POST") Post> findPostsByCategory(@PathVariable UUID categoryId, @RequestParam(defaultValue = "0") Integer page, @RequestParam(defaultValue = "10") Integer size) {
        return repository.findAllByCategoryId(PageRequest.of(page, size), categoryId, DEFAULT_POST);
    }

    public static final PostFetcher DEFAULT_POST = PostFetcher.$
            .allScalarFields()
            .content(false)
            .category(
                    CategoryFetcher.$
                            .allScalarFields()
            );
    public static final PostFetcher FULL_POST = PostFetcher.$
            .allScalarFields()
            .category(
                    CategoryFetcher.$
                            .allScalarFields()
            );
}

File: website/src/__generated/services/PostController.ts
import type { Executor } from '../';
import type { PostDto } from '../model/dto';
import type { Page, PostInput } from '../model/static';

export class PostController {
    
    constructor(private executor: Executor) {}
    
    async createPost(options: PostControllerOptions['createPost']): Promise<void> {
        let _uri = '/posts/';
        return (await this.executor({uri: _uri, method: 'PUT', body: options.body})) as void
    }
    
    async findPost(options: PostControllerOptions['findPost']): Promise<
        PostDto['PostController/FULL_POST']
    > {
        let _uri = '/posts/';
        _uri += encodeURIComponent(options.id);
        return (await this.executor({uri: _uri, method: 'GET'})) as PostDto['PostController/FULL_POST']
    }
    
    async findPosts(options: PostControllerOptions['findPosts']): Promise<
        Page<PostDto['PostController/DEFAULT_POST']>
    > {
        let _uri = '/posts/';
        let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
        let _value: any = undefined;
        _value = options.page;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'page='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.size;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'size='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        return (await this.executor({uri: _uri, method: 'GET'})) as Page<PostDto['PostController/DEFAULT_POST']>
    }
    
    async findPostsByCategory(options: PostControllerOptions['findPostsByCategory']): Promise<
        Page<PostDto['PostController/DEFAULT_POST']>
    > {
        let _uri = '/categories/';
        _uri += encodeURIComponent(options.categoryId);
        _uri += '/posts/';
        let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
        let _value: any = undefined;
        _value = options.page;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'page='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.size;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'size='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        return (await this.executor({uri: _uri, method: 'GET'})) as Page<PostDto['PostController/DEFAULT_POST']>
    }
}

export type PostControllerOptions = {
    'createPost': {readonly body: PostInput},
    'findPost': {readonly id: string},
    'findPosts': {readonly page?: number, readonly size?: number},
    'findPostsByCategory': {
        readonly categoryId: string, 
        readonly page?: number, 
        readonly size?: number
    }
}